File: app/src/main/java/net/veldor/flibustaloader/ui/BlacklistActivity.java
package net.veldor.flibustaloader.ui;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.RadioGroup;
import android.widget.Switch;
import android.widget.Toast;

import androidx.lifecycle.LiveData;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.R;
import net.veldor.flibustaloader.adapters.BlacklistAdapter;
import net.veldor.flibustaloader.selections.BlacklistItem;
import net.veldor.flibustaloader.utils.BlacklistAuthors;
import net.veldor.flibustaloader.utils.BlacklistBooks;
import net.veldor.flibustaloader.utils.BlacklistGenres;
import net.veldor.flibustaloader.utils.BlacklistSequences;
import net.veldor.flibustaloader.utils.MyPreferences;

import java.util.ArrayList;

public class BlacklistActivity extends BaseActivity {

    private RadioGroup mRadioContainer;
    private BlacklistAdapter mBlacklistAdapter;
    private RecyclerView mRecycler;
    private BlacklistBooks mBooksBlacklistContainer;
    private EditText mInput;
    private BlacklistAuthors mAuthorsBlacklistContainer;
    private BlacklistSequences mSequencesBlacklistContainer;
    private BlacklistGenres mGenresBlacklistContainer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.new_blacklist_activity);

        setupUi();
        setupObservers();
    }

    private void setupUi() {
        setupInterface();

        // скрою переход на данное активити
        Menu menuNav = mNavigationView.getMenu();
        MenuItem item = menuNav.findItem(R.id.goToBlacklist);
        item.setEnabled(false);
        item.setChecked(true);
        // добавлю идентификатор строки поиска
        mInput = findViewById(R.id.blacklist_name);
        mRecycler = findViewById(R.id.resultsList);
        mRecycler.setLayoutManager(new LinearLayoutManager(this));

        mBooksBlacklistContainer = App.getInstance().getBooksBlacklist();
        mAuthorsBlacklistContainer = App.getInstance().getAuthorsBlacklist();
        mSequencesBlacklistContainer = App.getInstance().getSequencesBlacklist();
        mGenresBlacklistContainer = App.getInstance().getGenresBlacklist();
        showBooks();
        // отслежу переключение типа добавления
        mRadioContainer = findViewById(R.id.blacklist_type);
        if (mRadioContainer != null) {
            mRadioContainer.setOnCheckedChangeListener((group, checkedId) -> {
                switch (checkedId) {
                    case R.id.blacklistBook:
                        showBooks();
                        break;
                    case R.id.blacklistAuthor:
                        showAuthors();
                        break;
                    case R.id.blacklistSequence:
                        showSequences();
                        break;
                    case R.id.blacklistGenre:
                        showGenres();
                        break;
                }
            });
        }

        Button subscribeBtn = findViewById(R.id.add_to_blacklist_btn);
        if (subscribeBtn != null) {
            subscribeBtn.setOnClickListener(this::addToBlacklist);
        }

        Switch switchOnlyRussian = findViewById(R.id.switchOnlyRussian);
        switchOnlyRussian.setChecked(MyPreferences.getInstance().isOnlyRussian());
        switchOnlyRussian.setOnCheckedChangeListener((buttonView, isChecked) -> MyPreferences.getInstance().setOnlyRussian(isChecked));
    }

    public void addToBlacklist(View view) {
        String value = mInput.getText().toString().trim();
        if (!value.isEmpty()) {
            // добавлю подписку в зависимости от типа
            switch (mRadioContainer.getCheckedRadioButtonId()) {
                case R.id.blacklistBook:
                    mBooksBlacklistContainer.addValue(value);
                    break;
                case R.id.blacklistAuthor:
                    mAuthorsBlacklistContainer.addValue(value);
                    break;
                case R.id.blacklistSequence:
                    mSequencesBlacklistContainer.addValue(value);
                    break;
                case R.id.blacklistGenre:
                    mGenresBlacklistContainer.addValue(value);
                    break;

            }
            mInput.setText("");
            Toast.makeText(this, "Добавляю значение " + value, Toast.LENGTH_LONG).show();
        } else {
            Toast.makeText(this, "Введите значение", Toast.LENGTH_LONG).show();
            mInput.requestFocus();
        }
    }

    private void showBooks() {
        // получу подписки на книги
        ArrayList<BlacklistItem> autocompleteValues = mBooksBlacklistContainer.getBlacklist();
        mBlacklistAdapter = new BlacklistAdapter(autocompleteValues);
        mRecycler.setAdapter(mBlacklistAdapter);
    }

    private void showAuthors() {
        // получу подписки на авторов
        ArrayList<BlacklistItem> autocompleteValues = mAuthorsBlacklistContainer.getBlacklist();
        mBlacklistAdapter = new BlacklistAdapter(autocompleteValues);
        mRecycler.setAdapter(mBlacklistAdapter);
    }

    private void showSequences() {
        // получу чёрный список серий
        ArrayList<BlacklistItem> autocompleteValues = mSequencesBlacklistContainer.getBlacklist();
        mBlacklistAdapter = new BlacklistAdapter(autocompleteValues);
        mRecycler.setAdapter(mBlacklistAdapter);
    }
    private void showGenres() {
        // получу чёрный список жанров
        ArrayList<BlacklistItem> autocompleteValues = mGenresBlacklistContainer.getBlacklist();
        mBlacklistAdapter = new BlacklistAdapter(autocompleteValues);
        mRecycler.setAdapter(mBlacklistAdapter);
    }

    @Override
    protected void setupObservers() {
        super.setupObservers();
        Log.d("surprise", "BlacklistActivity.java 95 setupObservers: setting up observers");
        // буду отслеживать изменения списка книг
        LiveData<Boolean> refresh = BlacklistBooks.mListRefreshed;
        refresh.observe(this, aBoolean -> {
            if (aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.blacklistBook) {
                refreshBooksBlacklist();
            }
        });


        // буду отслеживать изменения списка авторов
        LiveData<Boolean> authorRefresh = BlacklistAuthors.mListRefreshed;
        authorRefresh.observe(this, aBoolean -> {
            if (aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.blacklistAuthor) {
                refreshAuthorBlacklist();
            }
        });

        // буду отслеживать изменения списка серий
        LiveData<Boolean> sequencesRefresh = BlacklistSequences.mListRefreshed;
        sequencesRefresh.observe(this, aBoolean -> {
            if (aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.blacklistSequence) {
                refreshSequencesBlacklist();
            }
        });

        // буду отслеживать изменения списка жанров
        LiveData<Boolean> genresRefresh = BlacklistGenres.mListRefreshed;
        genresRefresh.observe(this, aBoolean -> {
            if (aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.blacklistGenre) {
                refreshGenresBlacklist();
            }
        });
    }


    private void refreshBooksBlacklist() {
        ArrayList<BlacklistItem> autocompleteValues = mBooksBlacklistContainer.getBlacklist();
        mBlacklistAdapter.changeList(autocompleteValues);
        mBlacklistAdapter.notifyDataSetChanged();
    }

    private void refreshAuthorBlacklist() {
        ArrayList<BlacklistItem> blacklist = mAuthorsBlacklistContainer.getBlacklist();
        mBlacklistAdapter.changeList(blacklist);
        mBlacklistAdapter.notifyDataSetChanged();
    }

    private void refreshSequencesBlacklist() {
        ArrayList<BlacklistItem> blacklist = mSequencesBlacklistContainer.getBlacklist();
        mBlacklistAdapter.changeList(blacklist);
        mBlacklistAdapter.notifyDataSetChanged();
    }

    private void refreshGenresBlacklist() {
        ArrayList<BlacklistItem> blacklist = mGenresBlacklistContainer.getBlacklist();
        mBlacklistAdapter.changeList(blacklist);
        mBlacklistAdapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/net/veldor/flibustaloader/ui/BlacklistActivity.kt
package net.veldor.flibustaloader.ui

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.inputmethod.EditorInfo
import android.widget.CompoundButton
import android.widget.RadioGroup
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.LinearLayoutManager
import co.mobiwise.materialintro.shape.Focus
import co.mobiwise.materialintro.shape.FocusGravity
import co.mobiwise.materialintro.shape.ShapeType
import co.mobiwise.materialintro.view.MaterialIntroView
import net.veldor.flibustaloader.R
import net.veldor.flibustaloader.adapters.BlacklistAdapter
import net.veldor.flibustaloader.databinding.ActivityBlacklistBinding
import net.veldor.flibustaloader.utils.*


class BlacklistActivity : BaseActivity() {
    lateinit var binding: ActivityBlacklistBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBlacklistBinding.inflate(layoutInflater)
        setContentView(binding.drawerLayout)
        setupUi()
        setupObservers()
    }

    private fun setupUi() {
        setupInterface()
        if (PreferencesHandler.instance.isEInk) {
            binding.blacklistBook.setTextColor(
                ResourcesCompat.getColor(
                    resources,
                    R.color.black,
                    null
                )
            )
            binding.blacklistAuthor.setTextColor(
                ResourcesCompat.getColor(
                    resources,
                    R.color.black,
                    null
                )
            )
            binding.blacklistSequence.setTextColor(
                ResourcesCompat.getColor(
                    resources,
                    R.color.black,
                    null
                )
            )
            binding.blacklistGenre.setTextColor(
                ResourcesCompat.getColor(
                    resources,
                    R.color.black,
                    null
                )
            )
            paintToolbar(binding.toolbar)
        }
        showHints()
        // скрою переход на данное активити
        val menuNav = mNavigationView.menu
        val item = menuNav.findItem(R.id.goToBlacklist)
        item.isEnabled = false
        item.isChecked = true
        // добавлю идентификатор строки поиска
        binding.resultsList.layoutManager = LinearLayoutManager(this)
        binding.resultsList.adapter = BlacklistAdapter(BlacklistBooks.instance.getBlacklist())
        binding.blacklistType.setOnCheckedChangeListener { _: RadioGroup?, checkedId: Int ->
            when (checkedId) {
                R.id.blacklistBook -> {
                    (binding.resultsList.adapter as BlacklistAdapter).changeList(BlacklistBooks.instance.getBlacklist())
                }
                R.id.blacklistAuthor -> {
                    (binding.resultsList.adapter as BlacklistAdapter).changeList(BlacklistAuthors.instance.getBlacklist())
                }
                R.id.blacklistSequence -> {
                    (binding.resultsList.adapter as BlacklistAdapter).changeList(BlacklistSequences.instance.getBlacklist())
                }
                R.id.blacklistGenre -> {
                    (binding.resultsList.adapter as BlacklistAdapter).changeList(BlacklistGenres.instance.getBlacklist())
                }
            }
        }
        binding.addToBlacklistBtn.setOnClickListener { addToBlacklist() }
        binding.switchOnlyRussian.isChecked = PreferencesHandler.instance.isOnlyRussian
        binding.switchOnlyRussian.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            PreferencesHandler.instance.isOnlyRussian = isChecked
        }

        binding.blacklistItemInput.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                // Do whatever you want here
                addToBlacklist()
                true
            } else false
        }
    }

    private fun showHints() {
        showFirstHelp()
    }

    private fun addToBlacklist() {
        val value = binding.blacklistItemInput.text.toString().trim { it <= ' ' }
        if (value.isNotEmpty()) {
            // добавлю подписку в зависимости от типа
            when (binding.blacklistType.checkedRadioButtonId) {
                R.id.blacklistBook -> {
                    BlacklistBooks.instance.addValue(value)
                }
                R.id.blacklistAuthor -> {
                    BlacklistAuthors.instance.addValue(value)
                }
                R.id.blacklistSequence -> {
                    BlacklistSequences.instance.addValue(value)
                }
                R.id.blacklistGenre -> {
                    BlacklistGenres.instance.addValue(value)
                }
            }
            binding.blacklistItemInput.setText("")
            Toast.makeText(this, "Добавляю значение $value", Toast.LENGTH_LONG).show()
        } else {
            Toast.makeText(this, "Введите значение", Toast.LENGTH_LONG).show()
            binding.blacklistItemInput.requestFocus()
        }
    }

    override fun setupObservers() {
        super.setupObservers()
        // буду отслеживать изменения списка книг
        BlacklistBooks.instance.liveBlacklistAdd.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistBook) {
                (binding.resultsList.adapter as BlacklistAdapter).itemAdded(it)
            }
        })
        BlacklistBooks.instance.liveBlacklistRemove.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistBook) {
                (binding.resultsList.adapter as BlacklistAdapter).itemRemoved(it)
            }
        })
        // буду отслеживать изменения списка книг
        BlacklistAuthors.instance.liveBlacklistAdd.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistAuthor) {
                (binding.resultsList.adapter as BlacklistAdapter).itemAdded(it)
            }
        })
        BlacklistAuthors.instance.liveBlacklistRemove.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistAuthor) {
                (binding.resultsList.adapter as BlacklistAdapter).itemRemoved(it)
            }
        })
        // буду отслеживать изменения списка книг
        BlacklistGenres.instance.liveBlacklistAdd.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistGenre) {
                (binding.resultsList.adapter as BlacklistAdapter).itemAdded(it)
            }
        })
        BlacklistGenres.instance.liveBlacklistRemove.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistGenre) {
                (binding.resultsList.adapter as BlacklistAdapter).itemRemoved(it)
            }
        })
        // буду отслеживать изменения списка книг
        BlacklistSequences.instance.liveBlacklistAdd.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistSequence) {
                (binding.resultsList.adapter as BlacklistAdapter).itemAdded(it)
            }
        })
        BlacklistSequences.instance.liveBlacklistRemove.observe(this, {
            if (binding.blacklistType.checkedRadioButtonId == R.id.blacklistSequence) {
                (binding.resultsList.adapter as BlacklistAdapter).itemRemoved(it)
            }
        })
    }
}

private fun BlacklistActivity.showFirstHelp() {
    Handler(Looper.getMainLooper()).postDelayed({
        kotlin.run {
            MaterialIntroView.Builder(this)
                .enableDotAnimation(true)
                .enableIcon(false)
                .setFocusGravity(FocusGravity.CENTER)
                .setFocusType(Focus.MINIMUM)
                .setDelayMillis(300)
                .setUsageId("blacklist type select")
                .enableFadeAnimation(true)
                .performClick(true)
                .setListener {
                    showSecondHelp()
                }
                .setInfoText(getString(R.string.blacklist_first_help_text))
                .setTarget(binding.blacklistBook)
                .setShape(ShapeType.CIRCLE)
                .show()
        }
    }, 100)
}

private fun BlacklistActivity.showSecondHelp() {
    Handler(Looper.getMainLooper()).postDelayed({
        kotlin.run {
            MaterialIntroView.Builder(this)
                .enableDotAnimation(true)
                .enableIcon(false)
                .setFocusGravity(FocusGravity.CENTER)
                .setFocusType(Focus.MINIMUM)
                .setDelayMillis(700)
                .setUsageId("blacklist value enter")
                .enableFadeAnimation(true)
                .performClick(true)
                .setListener {

                }
                .setInfoText(getString(R.string.blacklist_second_hint_text))
                .setTarget(binding.blacklistItemInput)
                .setShape(ShapeType.CIRCLE)
                .show()
        }
    }, 100)
}